src/crystal/compiler_rt/udivmodti4.cr

```diff
@@ -0,0 +1,132 @@
+# Function return the remainder of the unsigned division with overflow eg. `a % b`
+
+fun __udivmodti4(a : UInt128, b : UInt128, rem : UInt128*) : UInt128
+  n_udword_bits = sizeof(Int64) &* sizeof(Char)
+  n_utword_bits = sizeof(Int128) &* sizeof(Char)
+  n = a.unsafe_as(CompilerRT::UI128)
+  d = b.unsafe_as(CompilerRT::UI128)
+  q = CompilerRT::UI128.new
+  r = CompilerRT::UI128.new
+  sr = 0_u32
+
+  if n.info.high == 0
+    if d.info.high == 0
+      if rem
+        rem.value = (n.info.low % d.info.low).to_u128
+      end
+      n.info.low = n.info.low // d.info.low
+      return n.all
+    end
+    rem.value = n.info.low.to_u128 if rem
+    return 0_u128
+  end
+
+  if d.info.low == 0
+    if d.info.high == 0
+      if rem
+        n.info.high = n.info.high % d.info.low
+        rem.value = n.all
+      end
+      n.info.high = n.info.high // d.info.low
+      return n.all
+    end
+
+    if n.info.low == 0
+      if rem
+        r.info.high = n.info.high % d.info.high
+        r.info.low = 0
+        rem.value = r.all
+      end
+      n.info.high = n.info.high // d.info.high
+      return n.all
+    end
+
+    if (d.info.high & (d.info.high &- 1)) == 0 # if d is a power of 2
+      if rem
+        r.info.low = n.info.low
+        r.info.high = n.info.high & (d.info.high &- 1)
+        rem.value = r.all
+      end
+      n.info.high = n.info.high >> d.info.high.trailing_zeros_count
+      return n.all
+    end
+
+    sr = d.info.high.trailing_zeros_count &- n.info.high.trailing_zeros_count
```

Comment:
Should this be leading instead of trailing zeros?
https://github.com/llvm-mirror/compiler-rt/blob/master/lib/builtins/udivmodti4.c#L85

```
Built-in Function: int __builtin_clz (unsigned int x)
Returns the number of leading 0-bits in x, starting at the most significant bit position. If x is 0, the result is undefined.

Built-in Function: int __builtin_ctz (unsigned int x)
Returns the number of trailing 0-bits in x, starting at the least significant bit position. If x is 0, the result is undefined.
```

Reply:
There are more mentions of `clz` in https://github.com/llvm-mirror/compiler-rt/blob/master/lib/builtins/udivmodti4.c#L85 but no leading_zeros_count here.

Reply:
Thanks I took a journey down llvm MathExtras but it looks like the Crystals Int `trailing_zeros_count` and `leading_zeros_count` works.  I have updated udivmodti4 if you want to take another look.